File: src/main/java/dev/psygamer/wireframe/lib/block/BlockProperty.java
package dev.psygamer.wireframe.lib.block;

import net.minecraft.state.Property;
import net.minecraftforge.client.model.generators.ConfiguredModel;

import java.util.Optional;

public abstract class BlockProperty <T extends Comparable<T>> extends Property<T> {
	
	protected BlockProperty(final String name, final Class<T> valueClass) {
		super(name, valueClass);
	}
	
	public abstract ConfiguredModel.Builder<?> applyBlockModelModification(Comparable<? extends Comparable<?>> value, ConfiguredModel.Builder<?> modelBuilder);
	
	@Override
	public abstract String getName(final T value);
	
	@Override
	public abstract Optional<T> getValue(final String name);
}

File: src/main/kotlin/dev/psygamer/wireframe/api/block/BlockProperty.kt
package dev.psygamer.wireframe.api.block

import java.util.*
import dev.psygamer.wireframe.nativeapi.block.NativeBlockProperty

abstract class BlockProperty<T : Comparable<T>>(propertyName: String, defaultValue: T) {

	val propertyName: String
	val defaultValue: T

	internal var mcNative: net.minecraft.state.Property<T>

	init {
		Objects.requireNonNull(propertyName, "The property name may not be null")

		this.propertyName = propertyName
		this.defaultValue = defaultValue

		this.mcNative = NativeBlockProperty(this)
	}

	abstract fun getValue(valueName: String): Optional<T>
	abstract fun getValueName(value: T): String

	abstract val possibleValues: Set<T>
}
